### ISSUE TITLE
Realm on async thread not notify instances of Realm on main looper

### ISSUE THREAD

**Issue by ppamorim** [realm/realm-java#4106]
I have an application that updates a value of a determined model. Imagine this code, this is very simple:

This code runs very fast, I think it is a problem of synchronisation.

```kotlin
override fun runAsync(id : Int, duration: Long) { // Duration is 1000
    try {

      if (id > -1) {
        throw IllegalStateException("id is invalid, impossible to update duration")
      }

      //INVALID_VALUE is -1
      if (duration <= INVALID_VALUE) {
        throw IllegalStateException("Duration is a invalid value, aborting")
      }

      val realm = Realm.getDefaultInstance()
      realm.executeTransaction {
        val model = realm.where(Model::class.java)
            ?.equalTo("id", id)
            ?.findFirst()

        model?.let { model ->
          if (model.layers.isNotEmpty) { //Always true
            model.layers.last().duration = duration
          }
        }

      }
      realm.close()

      //Realm used just to assert that the update is correct.
      val tempRealm = Realm.getDefaultInstance()
      val model = tempRealm.where(Model::class.java)
          ?.equalTo("id", id)
          ?.findFirst()

      model?.layers?.last()?.let {
        println("Value updated: ${it.duration}") //Return 1000
      }

      tempRealm.close()

      //Notify the main thread to get the data
      notifyMainThread()

    } catch (e: Exception) {
      e.printStackTrace()
    }
  }
```

No exception was throw, no error. Everything is ok.

After the app return to the main looper, I did query for this Model again, using the same id.

```kotlin
//After return to the main looper
fun notifyMainThread() {
  val realm = Realm.getDefaultInstance()
  val model = tempRealm.where(Model::class.java)
          ?.equalTo("id", id)
          ?.findFirst()

      model?.layers?.last()?.let {
        println("Value updated: ${it.duration}") //Return INVALID_VALUE
      }
}
```

I added a `RealmChangeListener` and it solved the problem, but I think that is not correct, I am using a async task to avoid any kinda of problem on inserting and updating. Is there any way to force the async task to be locked until the notify of data changed on other Realm?

Realm: 2.3.0
Kotlin: 1.0.4

Thank you.

**Comment by Zhuinden** [comment]
Async callbacks are not called if the associated Realm instance is closed.

**Comment by Zhuinden** [comment]
I think this is caused by https://github.com/realm/realm-java/issues/3427

**Comment by ppamorim** [comment]
@Zhuinden Sorry I missed to app the close on this component, but the problem wasn't fixed.

**Comment by Zhuinden** [comment]
Technically it is not guaranteed that the Realm on the UI thread is **immediately updated on the next event loop** after an asynchronous write on another thread, so AsyncTask returns "too quickly" and the Realm isn't updated yet.

I have not figured out a solution to this problem, but maybe @cmelchior knows.

The recommendation is to use RealmChangeListener instead of direct UI-thread callbacks.

**Comment by ppamorim** [comment]
@Zhuinden Sometimes this value comes correct, so I only add the listener if needed.

**Comment by ppamorim** [comment]
Make the method `Realm.refresh()` great again

**Comment by Zhuinden** [comment]
Make AsyncTasks great again

**Comment by ppamorim** [comment]
@beeender By some Realm some instances of Realm are not being updated (more than a second of difference between the start of this Realm instantiation and the usage).
I would like to know if I can force the refresh of the Realm, not using any cache (of the method `getDefaultInstance`). By some reason, sometimes when I start a new Realm inside a async thread, this Realm is not synchronised with the latest Realm update. That is crazy...
So... I would use the method `waitForChange()` but I don't know if this Realm is waiting for changes or not.

**Comment by ppamorim** [comment]
I can't use the change listener, this is not the main looper...

**Comment by Zhuinden** [comment]
@ppamorim this code refreshes the Realm on the current thread: http://stackoverflow.com/a/38839808/2413303

**Comment by ppamorim** [comment]
@Zhuinden Thank you! I will try it

**Comment by ppamorim** [comment]
@Zhuinden Unfortunately I can't use the variable `realm.handlerController`.

**Comment by ppamorim** [comment]
I really don't mind to recreate the instance of Realm, this is not a problem on this part of my source code. Any alternative?

**Comment by Zhuinden** [comment]
You can, just pay attention to the package in that answer

**Comment by ppamorim** [comment]
I am using Gradle I don't know if I can modify it.

**Comment by Zhuinden** [comment]
You just need to create a package in the root called `io` and inside it create a package called `realm` and from there you can create RealmRefresh which can access the handler controller

**Comment by ppamorim** [comment]
Very good way to "inject" code.

**Comment by ppamorim** [comment]
@Zhuinden You saved my life! Thank you very much!

**Comment by Zhuinden** [comment]
Glad to hear it worked :)
I posted this issue https://github.com/realm/realm-java/issues/3476 about this a long time ago but alas

**Comment by kneth** [comment]
I'll close the issue since @Zhuinden's suggestion seems to help.

**Comment by ppamorim** [comment]
@kneth Do you think that's a correct way to fix this problem? I don't think so...

**Comment by kneth** [comment]
@ppamorim I can reopen the issue if the issue hasn't been solved (and the stackoverflow link wasn't of any help to you).

**Comment by ppamorim** [comment]
@kneth I believe that you don't need to do it. I will be waiting the rewrite of the async queries and the implementation of a proper way to sync the `Realm`.

**Comment by Zhuinden** [comment]
Fixed by https://github.com/realm/realm-java/pull/4368 in Realm 3.1.1

**Comment by Zhuinden** [comment]
You can use [refresh()](https://github.com/realm/realm-java/pull/4515) method available since Realm 3.2+
